File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    queue = deque([(1,0,0)])
    ans = float("INF")
    DIR = [(0,1), (1,0), (-1,0), (0,-1)]
    visited = [[False for _ in range(len(maps[0]))] for _ in range(len(maps))]
    
    while queue:
        t_cnt,ty,tx = queue.popleft()
        if ty == len(maps)-1 and tx == len(maps[0])-1:
            ans = min(t_cnt, ans)
        else:
            for dy,dx in DIR:
                if 0 <= ty+dy < len(maps) and 0 <= tx + dx < len(maps[0]) and maps[ty+dy][tx+dx] == 1 and not visited[ty+dy][tx+dx]:
                    queue.append((t_cnt + 1, ty+dy, tx+dx))
                    visited[ty + dy][tx + dx] = True
    if ans == float("INF"):
        return -1
    else:
        return ans

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[] dy = {0,0,-1,1};
    static int[] dx = {-1,1,0,0};
    static int len_y, len_x;
    
    public int solution(int[][] maps) {
        int answer = Integer.MAX_VALUE;
        len_y = maps.length;
        len_x = maps[0].length;
        
        boolean[][] visited = new boolean[len_y][len_x];
        PriorityQueue<List<Integer>> queue = new PriorityQueue<>((o1,o2) -> o1.get(0) - o2.get(0));
        queue.add(List.of(1,0,0)); // wei, y, x
        visited[0][0] = true;
        
        while(!queue.isEmpty()){
            List<Integer> tmp = queue.poll();
            int wei = tmp.get(0);
            int y = tmp.get(1);
            int x = tmp.get(2);
            if(y == len_y-1 && x== len_x-1){
                answer = Math.min(answer, wei);
            }
            
            for(int i = 0; i < 4; i ++){
                if((0 <= y+dy[i]) && (y+dy[i] < len_y) && (0 <= x+dx[i]) && (x+dx[i] < len_x) && maps[y+dy[i]][x+dx[i]] == 1 && !visited[y+dy[i]][x+dx[i]]){
                    visited[y+dy[i]][x+dx[i]] = true;
                    queue.add(List.of(wei+1, y+dy[i], x+dx[i]));
                }
            }
        }
        if(answer == Integer.MAX_VALUE){
            return -1;
        }
        return answer;
    }
}